---
title: "Revision figures"
output: html_notebook
---

```{r}
rm(list = setdiff(ls(), "so_all"))
dir_proj <- "/singerlab/linglin/Th17_single_cell_eae_ut/"

suppressPackageStartupMessages(
  {
    library(Seurat)
    library(Matrix)
    library(ggplot2)
    library(grid)
    library(RColorBrewer)
    library(gridExtra)
    library(ggpubr)
    library(ggrepel)
    library(ggbeeswarm)
    library(dplyr)
    library(tidyr)
    library(tibble)
    library(openxlsx)
    library(pheatmap)
    library(stringr)
    source(paste0(dir_proj, "1_code/utils.R"))
  }
)

#### configuration ####
today <- Sys.Date()
dir_out <- paste0(dir_proj, "3_output/Revision_cell/")
if(!dir.exists(dir_out)) dir.create(dir_out, recursive = T)

tissue_vec_ut <- c("SPL", "MLN", "PP", "SI", "COL")
```


## 1. TCR-specificity of tissue Th17 cells by GLIPH
Analysis in 1_code/tcr/revision_specificity.Rmd
```{r}
two_beta <- readRDS(file = "../../2_pipeline/TCR/gliph/cells_two_beta.rds")
# res <- read.table("../../2_pipeline/TCR/gliph/2021-08-10_ut_eae_cluster.csv", sep = ",", header = T)
# res <- read.table("../../2_pipeline/TCR/gliph/2021-08-10_ut_valid_cluster.csv", sep = ",", header = T)
res <- read.table("../../2_pipeline/TCR/gliph/2021-09-22_ut_cluster.csv", sep = ",", header = T)

length(unique(res$index[grepl("global", res$type)]))

res <- res %>% 
  # filter((grepl("global", type) & (Fisher_score < 0.05)) | (pattern == "single"))  # only keep significant global clusters or singlets
  filter((grepl("global", type) & (Fisher_score < 0.05) & (pattern != "single")))  # only keep significant global clusters
length(unique(res$index[grepl("global", res$type)]))

two_beta_df <- Reduce(rbind, two_beta[grep("UT", names(two_beta))]) %>% 
  mutate(CDR3b_1 = sub(":.*", "", CDR3b),
         CDR3b_2 = sub(".*:", "", CDR3b))
two_beta_df <- unique(two_beta_df[,-1])

idx_1 <- which(two_beta_df$CDR3b_1 %in% res$TcRb) # which cells are found in a valid cluster based on the first beta chain
idx_2 <- which(two_beta_df$CDR3b_2 %in% res$TcRb) # which cells are found in a valid cluster based on the second beta chain
idx_1_2 <- intersect(idx_1, idx_2) # both chains are in cluster(s)

res$index_merge <- res$index
for (i in idx_1_2){
  res_idx_1 <- which(as.character(res$TcRb) == as.character(two_beta_df[i,"CDR3b_1"]))
  res_idx_2 <- which(as.character(res$TcRb) == as.character(two_beta_df[i,"CDR3b_2"]))
  cluster_idx_1 <- res$index[res_idx_1[1]]
  cluster_idx_2 <- res$index[res_idx_2[1]]
  cat("original cluster index for", i, ":", cluster_idx_1, cluster_idx_2, "\n")
  new_cluster_idx <- paste0(cluster_idx_1, "_m")
  res$index_merge[res$index == cluster_idx_1] <- new_cluster_idx
  res$index_merge[res$index == cluster_idx_2] <- new_cluster_idx
}
# table(res$index_merge)

mtx <- res %>% 
  mutate(tissue = sub(".*:", "", Sample),
         treatment = sub(".*?_", "", Sample) %>% sub("_.*", "", .),
         tt = paste(tissue, treatment, sep = "_"),
         index = paste0("Cluster_", index_merge)) %>% 
  # filter(tissue != "NA", treatment == "UT") %>%
  group_by(index, tissue) %>% 
  summarise(cluster_size = sum(Freq)) %>% 
  ungroup() %>% 
  spread(key = tissue, value = cluster_size) %>% 
  column_to_rownames(var = "index")
mtx[is.na(mtx)] <- 0
mtx <- mtx[,tissue_vec_ut]
cluster_size_vec <- rowSums(mtx)
summary(cluster_size_vec)
mtx <- mtx[cluster_size_vec > 0, ]
cluster_size_vec <- rowSums(mtx)
summary(cluster_size_vec)
pct_mtx <- mtx / rowSums(mtx) * 100

dim(mtx)
table(rowSums(mtx>0))

which_specific <- which((rowSums(mtx > 0)) == 1)
which_not_specific <- setdiff(1:nrow(mtx), which_specific)
idx_row_1 <- which_specific[do.call(order, -cbind(pct_mtx, cluster_size_vec)[which_specific, ])] 
idx_row_2 <- which_not_specific[do.call(order, -cbind(pct_mtx, cluster_size_vec)[which_not_specific,])]
idx_row <- c(idx_row_1, idx_row_2) 
pct_mtx <- pct_mtx[idx_row,]
p <- pheatmap(pct_mtx,
         color = colorRampPalette(c("white", "red"))(100),
         cluster_rows = F,
         cluster_cols = F,
         clustering_method = "complete",
         show_rownames = F,
         silent = F, 
         # border_color = "grey95",
         main = "")
which_col_names <- which(p$gtable$layout$name == "col_names")
p$gtable$grobs[[which_col_names]]$rot = 0
p$gtable$grobs[[which_col_names]]$hjust = 0.5
p$gtable$grobs[[which_col_names]]$vjust = 1
p$gtable$grobs[[which_col_names]]$gp$fontsize = 10

# p_total <- pheatmap(matrix(cluster_size_vec[idx_row], ncol = 1, dimnames = list(c(), c("Total"))),
#          color = colorRampPalette(c("grey90", "black"))(10),
#          breaks = seq(0, 10, 1),
#          legend_breaks = c(0, 2, 4, 6, 8, 10),
#          legend_labels = c("0", "2", "4", "6", "8", "10"),
#          cluster_rows = F,
#          cluster_cols = F,
#          clustering_method = "complete",
#          show_rownames = F,
#          # legend = F,
#          silent = F, 
#          border_color = "grey95",
#          main = "")
# which_col_names <- which(p_total$gtable$layout$name == "col_names")
# p_total$gtable$grobs[[which_col_names]]$rot = 0
# p_total$gtable$grobs[[which_col_names]]$hjust = 0.5
# p_total$gtable$grobs[[which_col_names]]$vjust = 1
# p_total$gtable$grobs[[which_col_names]]$gp$fontsize = 13
# 
# 
# pdf(paste0(dir_out, "Fig1_heatmap_GLIPH_cluster_by_tissue_UT_", today, ".pdf"), width = 6, height = 4)
# grid.arrange(p$gtable, p_total$gtable, nrow = 1, widths = c(4, 1))
# dev.off()
pdf(paste0(dir_out, "Fig1_heatmap_GLIPH_cluster_by_tissue_UT_no_total_", today, ".pdf"), width = 2.3, height = 5)
grid.arrange(p$gtable)
dev.off()
```


## 2. Source of IL-23

(1) SPL sc data set

Waiting for data.

(2) Smillie UC data set.

<!-- Preprocessing see ../other_data/human_smillie_preprocessing.R -->
```{r}
dat <- Matrix::readMM("../../0_data/other/Smillie_human_UC/SCP259/expression/5cdc540d328cee7a2efc234a/gene_sorted-Imm.matrix.mtx")
dat_gene <- read.table("../../0_data/other/Smillie_human_UC/SCP259/expression/5cdc540d328cee7a2efc234a/Imm.genes.tsv", stringsAsFactors = F)
dat_barcode <- read.table("../../0_data/other/Smillie_human_UC/SCP259/expression/5cdc540d328cee7a2efc234a/Imm.barcodes2.tsv", stringsAsFactors = F)
all_meta <- read.table("../../0_data/other/Smillie_human_UC/SCP259/metadata/all.meta2.txt", stringsAsFactors = F, sep = "\t", header = T)

rownames(dat) <- dat_gene$V1
colnames(dat) <- dat_barcode$V1
dat_meta <- all_meta %>% 
  filter(NAME %in% dat_barcode$V1) %>% 
  filter(Health == "Healthy") %>% 
  column_to_rownames(var = "NAME") 

so <- CreateSeuratObject(counts = dat[,rownames(dat_meta)], meta.data = dat_meta)
rm(dat, dat_gene, dat_barcode, all_meta)

table(so$Cluster, useNA = "ifany")

c("IL23A", "IL12B") %in% rownames(so[["RNA"]]@counts)

# rownames(so[["RNA"]]@counts)[grep("IL12", rownames(so[["RNA"]]@counts))] #"IL12A"   "IL12RB1" "IL12RB2"

# barplot of percent cell expressing IL23A and IL12B
plt_df <- data.frame(
  IL23A = so[["RNA"]]@data["IL23A",],
  cluster = so$Cluster,
  subject = so$Subject
) %>% group_by(cluster, subject, .drop = FALSE) %>% 
  summarise(pct_IL23A = 100 * mean(IL23A > 0)) %>% 
  ungroup() %>% 
  group_by(cluster, .drop = FALSE) %>% 
  summarise(avg_pct_IL23A = mean(pct_IL23A, na.rm = T),
            hi_pct_IL23A = avg_pct_IL23A + sd(pct_IL23A, na.rm = T) / sqrt(sum(!is.na(pct_IL23A))),
            lo_pct_IL23A = avg_pct_IL23A - sd(pct_IL23A, na.rm = T) / sqrt(sum(!is.na(pct_IL23A))))

p_il23a <- ggplot(plt_df) +
  geom_bar(aes(x = cluster, y = avg_pct_IL23A), 
           stat = "identity", width = 0.6, color = "black", fill = "#808080", size = 0.5) +
  geom_errorbar(aes(x = cluster, ymin = lo_pct_IL23A, ymax = hi_pct_IL23A), width = 0.3) +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = "", y = "Percent of cells") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 8, color = "black"),
        plot.margin = unit(c(1,1,1,1), "cm"))
pdf(paste0(dir_out, "Fig2_barplot_IL23A_Smillie_", today, ".pdf"), width = 7, height = 5)
p_il23a
dev.off()
```




(3) Wheeler & Quintana data. Preprocessing see 1_code/
```{r}
dir_wheeler <- "../../2_pipeline/cxcl16/2020-02-26/"
so <- readRDS(paste0(dir_wheeler, "so_combined.rds"))

### annotate clusters
## quantile normalization
library(preprocessCore)
annot_names <- colnames(so@meta.data)[grep("Cluster_", colnames(so@meta.data))]
scores_q <- normalize.quantiles(as.matrix(so@meta.data[,annot_names]))
so$annotation_score <- factor(annot_names[apply(scores_q, 1, which.max)] %>% sub("Cluster_", "C", .), levels = sub("Cluster_", "C", annot_names))
pheatmap(table(so$seurat_clusters, so$annotation_score) / as.numeric(table(so$seurat_clusters)) * 100,
         color = colorRampPalette(c("white", "red"))(100),
         breaks = seq(0, 100, length.out = 101),
         cluster_rows = F, cluster_cols = F,
         display_numbers = T, number_format = "%.1f", number_color = "black", fontsize_number = 7)

sample_info <- c(
"GSM3717013", "Priming_replicate1",
"GSM3717014", "Priming_replicate2",
"GSM3717015", "Priming_replicate3",
"GSM3717016", "CFA_replicate1",
"GSM3717017", "Acute_replicate1",
"GSM3717018", "Acute_replicate2",
"GSM3717019", "Acute_replicate3",
"GSM3717020", "Naive_replicate1",
"GSM3717021", "Naive_replicate2",
"GSM3717022", "Remitting_replicate1",
"GSM3717023", "Remitting_replicate2",
"GSM3717024", "Remitting_replicate3",
"GSM3717025", "CFA_replicate2",
"GSM3717026", "CFA_replicate3",
"GSM3717027", "Naive_replicate3",
"GSM4002707", "Acute_replicate4",
"GSM4002708", "Acute_replicate5",
"GSM4002709", "Acute_replicate6",
"GSM4002710", "Naive_replicate4",
"GSM4002711", "Remitting_replicate4",
"GSM4002712", "Remitting_replicate5",
"GSM4002713", "Remitting_replicate6",
"GSM4002714", "Naive_replicate5",
"GSM4002715", "Priming_replicate4",
"GSM4002716", "Priming_replicate5",
"GSM4002717", "Priming_replicate6",
"GSM4002718", "Naive_replicate6"
) %>% matrix(ncol = 2, byrow = T) %>% data.frame() %>% 
  select(GSM = X1, sample_name = X2) %>% 
  mutate(treatment = gsub( "_.*$", "",  sample_name))

so$sample_name <- sample_info$sample_name[match(so$orig.ident, sample_info$GSM)]
so$condition <-  factor(gsub( "_.*$", "",  so$sample_name), levels = c("Naive", "Priming", "CFA", "Acute", "Remitting"))

c("Il23a", "Il12b") %in% rownames(so[["integrated"]]@data)
c("Il23a", "Il12b") %in% rownames(so[["RNA"]]@data)

# DefaultAssay(so) <- "integrated"
# FeaturePlot(so, features = "Il12b", min.cutoff = "q5", sort.cell = T)
DefaultAssay(so) <- "RNA"
FeaturePlot(so, features = "Il23a", min.cutoff = "q5", sort.cell = T)
FeaturePlot(so, features = "Il12b", min.cutoff = "q5", sort.cell = T)


plt_df <- data.frame(
  Il23a = so[["RNA"]]@data["Il23a",],
  Il12b = so[["RNA"]]@data["Il12b",],
  condition = so$condition,
  cluster = so$annotation_score,
  subject = so$sample_name
) %>% filter(condition == "Priming") %>% 
  group_by(cluster, subject, .drop = FALSE) %>% 
  summarise(pct_Il23a = 100 * mean(Il23a > 0),
         pct_Il12b = 100 * mean(Il12b > 0)) %>% 
  ungroup() %>% 
  group_by(cluster, .drop = FALSE) %>% 
  summarise(avg_pct_Il23a = mean(pct_Il23a, na.rm = T),
            hi_pct_Il23a = avg_pct_Il23a + sd(pct_Il23a, na.rm = T) / sqrt(sum(!is.na(pct_Il23a))),
            lo_pct_Il23a = pmax(avg_pct_Il23a - sd(pct_Il23a, na.rm = T) / sqrt(sum(!is.na(pct_Il23a))), 0),
            avg_pct_Il12b = mean(pct_Il12b, na.rm = T),
            hi_pct_Il12b = avg_pct_Il12b + sd(pct_Il12b, na.rm = T) / sqrt(sum(!is.na(pct_Il12b))),
            lo_pct_Il12b = pmax(avg_pct_Il12b - sd(pct_Il12b, na.rm = T) / sqrt(sum(!is.na(pct_Il12b))), 0))

p_il23a <- ggplot(plt_df) +
  geom_bar(aes(x = cluster, y = avg_pct_Il23a), 
           stat = "identity", width = 0.6, color = "black", fill = "#808080", size = 0.5) +
  geom_errorbar(aes(x = cluster, ymin = lo_pct_Il23a, ymax = hi_pct_Il23a), width = 0.3) +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = "", y = "Percent of cells") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 8, color = "black"),
        plot.margin = unit(c(0.5, 0.5, 0.5, 0.5), "in"))

p_il12b <- ggplot(plt_df) +
  geom_bar(aes(x = cluster, y = avg_pct_Il12b), 
           stat = "identity", width = 0.6, color = "black", fill = "#808080", size = 0.5) +
  geom_errorbar(aes(x = cluster, ymin = lo_pct_Il12b, ymax = hi_pct_Il12b), width = 0.3) +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = "", y = "Percent of cells") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 8, color = "black"),
        plot.margin = unit(c(0.5, 0.5, 0.5, 0.5), "in"))


pdf(paste0(dir_out, "Fig2_barplot_Il23a_Quintana_", today, ".pdf"), width = 7.5, height = 5.5)
p_il23a
dev.off()
pdf(paste0(dir_out, "Fig2_barplot_Il12b_Quintana_", today, ".pdf"), width = 7.5, height = 5.5)
p_il12b
dev.off()

```
Shafflick data.
```{r}
# date_analysis <- "2020-07-01"
# meta_pbmc <- readRDS(file = paste0(dir_proj, "2_pipeline/human_shafflick/", date_analysis, "/meta_pbmc.rds"))
# meta_csf <- readRDS(file = paste0(dir_proj, "2_pipeline/human_shafflick/", date_analysis, "/meta_csf.rds"))
```




## 3. Role of IFNGR1 in driving pathogenicity of the CXCR6 population
Analysis in 1_code/ifngr1ko/analysis.Rmd

Heatmap of DE genes in CXCR6+ population.
```{r}
thresh_padj <- 0.15
date_de <- "2021-07-29"
res_edger <- readRDS(file = paste0("../../2_pipeline/Ifngr1ko/de_edger_CXCR6+_", date_de, ".rds"))
date_prep <- "2021-05-17"
dds <- readRDS(file = paste0("../../2_pipeline/Ifngr1ko/dds_", date_prep, ".rds"))
exclude_bad_samples <- c("D12", "E12", "F04") # based on alignment rate and total number reads
dds <- dds[,!(dds$well_parsed %in% exclude_bad_samples)]
vsd <- DESeq2::vst(dds)
plt_df <- as.data.frame(res_edger$table) %>%
    rownames_to_column(var = "gene") %>%
    filter(!is.na(PValue)) %>%
    mutate(up_in = ifelse(FDR >= thresh_padj, "NS", ifelse(logFC < 0, "WT", "KO")) %>% factor(., levels = c("KO", "WT", "NS")))
expr <- assay(vsd)[plt_df$gene[plt_df$up_in != "NS"], dds$cell_type == "CXCR6+"] %>% t %>% scale %>% t
hc_row <- hclust(dist(expr), method = "complete")
expr <- expr[hc_row$order,]
hc_col <- hclust(dist(t(expr)), method = "complete")
expr <- expr[,hc_col$order]

## heatmap
ann_col <- data.frame(
  Condition = toupper(as.character(colData(dds)[,"condition"]))
)
rownames(ann_col) <- colnames(dds)
ann_colors <- list(Condition = c("WT" = "black", "KO" = rgb(205, 110, 49, maxColorValue = 255)))
p <- pheatmap(expr,
         cluster_cols = F, cluster_rows = F,
         color = colorRampPalette(c("blue", "white", "red"))(100),
         breaks = seq(-2, 2, length.out = 101),
         annotation_col = ann_col,
         annotation_colors = ann_colors,
         show_rownames = T,
         show_colnames = F,
         main = "")
which_row_names <- which(p$gtable$layout$name == "row_names")
# p$gtable$grobs[[which_row_names]]$rot = 0
# p$gtable$grobs[[which_row_names]]$hjust = 0
# p$gtable$grobs[[which_row_names]]$vjust = 1
p$gtable$grobs[[which_row_names]]$gp$fontsize = 8
p$gtable$grobs[[which_row_names]]$gp$font = 3L

pdf(paste0(dir_out, "Fig3_heatmap_DEG_Ifngr1KO_", today, ".pdf"), width = 5, height = 7)
grid.arrange(p$gtable)
dev.off()
```


## 4. Differential expression analysis of ex- vs. current- Th17 cells
```{r}
tissue_vec <- c("SPL", "MLN", "PP", "SI", "COL")
res <- sapply(tissue_vec, simplify = F, FUN = function(tissue) {
  readRDS(paste0("../../2_pipeline/differential_expression/UT_GFPpos_vs_GFPneg/2021-08-31/", tissue, "_results.rds"))$table
})

library(openxlsx)
wb <- createWorkbook()
for (tissue in tissue_vec) {
  addWorksheet(wb, sheetName = tissue)
  writeData(wb, sheet = tissue, x = res[[tissue]], rowNames = T)
}
saveWorkbook(wb, file = paste0(dir_out, "DE_results_UT_intra_GFPpos_vs_neg_", today, ".xlsx"), overwrite = T)

thresh_fdr <- 0.05
thresh_lfc <- log2(1.5)
plt_ls <- lapply(res, function(x) {
  x %>% rownames_to_column(var = "gene") %>% 
    mutate(up_in = ifelse(FDR >= thresh_fdr, "ns", ifelse(logFC > thresh_lfc, "GFP+", ifelse(logFC < -thresh_lfc, "GFP-", "ns"))),
           rank_pval = rank(PValue), rank_lfc_inc = rank(logFC), rank_lfc_dec = rank(-abs(logFC)),
           lp = -log10(PValue)) %>% 
    arrange(-abs(logFC))
})

plist <- lapply(tissue_vec, function(x){
  plt_df <- plt_ls[[x]]
  if (sum(plt_df$up_in != "ns") >= 30) {
    plt_df <- plt_df %>% mutate(gene_label = ifelse(up_in != "ns" & (rank_pval < 10 | rank_lfc_inc < 10 | rank_lfc_dec < 10), gene, NA))
  } else {
    plt_df <- plt_df %>% mutate(gene_label = ifelse((up_in != "ns" | rank_pval < 10 | rank_lfc_inc < 10 | rank_lfc_dec < 10), gene, NA))
  }
  # plt_df$logFC[plt_df$gene %in% c("Il17a", "Il17a-GFP")] <- pmin(plt_df$logFC[plt_df$gene %in% c("Il17a", "Il17a-GFP")],
  #                                                                max(plt_df$logFC[!(plt_df$gene %in% c("Il17a", "Il17a-GFP"))]))
  # plt_df$PValue[plt_df$gene %in% c("Il17a", "Il17a-GFP")] <- pmax(plt_df$PValue[plt_df$gene %in% c("Il17a", "Il17a-GFP")],
  #                                                                 min(plt_df$PValue[!(plt_df$gene %in% c("Il17a", "Il17a-GFP"))]))
  # plt_df$FDR[plt_df$gene %in% c("Il17a", "Il17a-GFP")] <- pmax(plt_df$FDR[plt_df$gene %in% c("Il17a", "Il17a-GFP")],
  #                                                              min(plt_df$FDR[!(plt_df$gene %in% c("Il17a", "Il17a-GFP"))]))
  if (x == "SPL") {
    plt_df$lp <- pmin(plt_df$lp, 15)
    plt_df$logFC <- pmin(plt_df$logFC, 0.5)
  } else if (x == "MLN") {
    plt_df$lp <- pmin(plt_df$lp, 25)
    plt_df$logFC <- pmin(plt_df$logFC, 1.25)
  } else if (x == "PP") {
    plt_df$lp <- pmin(plt_df$lp, 200)
    plt_df$logFC <- pmin(plt_df$logFC, 2)
  } else if (x == "SI") {
    plt_df$lp <- pmin(plt_df$lp, 100)
    plt_df$logFC <- pmin(plt_df$logFC, 2)
  } else if (x == "COL") {
    plt_df$lp <- pmin(plt_df$lp, 100)
    plt_df$logFC <- pmin(plt_df$logFC, 1.75)
  } else {
    stop("Invalid tissue!!!")
  }
  
  p <- ggplot(plt_df, aes(x = logFC, y = lp, color = up_in, label = gene_label)) +
    geom_point(size = 0.5) +
    geom_text_repel(size = 2, segment.size = 0.1, seed = 1) +
    scale_color_manual(values = c("GFP+" = "darkgreen", "GFP-" = "grey20", "ns" = "grey80")) +
    scale_x_continuous(limits = c(-max(plt_df$logFC), max(plt_df$logFC)), expand = expansion(mult = c(0.01, 0.01))) +
    scale_y_continuous(limits = c(0, max(plt_df$lp)), expand = expansion(mult = c(0.05, 0.01))) +
    # labs(title = x, caption = paste0("FDR < ", thresh_fdr, ", fold change > ", sprintf("%.2f", 2^thresh_lfc))) +
    labs(title = x, color = "Up-regulated in", y = "-log10(p)", x = "Log2 fold change (GFP+ vs. GFP-)") +
    theme_bw() +
    theme(plot.title = element_text(size = 12, hjust = 0.5, face = "bold"),
          axis.title = element_text(size = 10, color = "black"))
  return(p)
})

library(gridExtra)
library(cowplot)
pdf(file = paste0(dir_out, "Fig4_volcano_GFP_pos_vs_neg_", today, ".pdf"), width = 10, height = 6)
grid.arrange(grobs = c(lapply(plist, function(x){x + theme(legend.position = "none")}), list(get_legend(plist[[3]]))), nrow = 2)
dev.off()

# pdf(file = paste0("2_pipeline/revision/volcano_ex_vs_current_1row_", today, ".pdf"), width = 12, height = 6)
# grid.arrange(grobs = lapply(plist, function(x){x + theme(legend.position = "none")}), nrow = 1,
#              right = get_legend(plist[[3]]))
# dev.off()
# paste0("FDR < ", thresh_fdr, ", fold change > ", sprintf("%.2f", 2^thresh_lfc))


library(limma)
deg_ls <- lapply(plt_ls, function(x){x$gene[x$up_in == "GFP+"]})
all_deg <- Reduce(union, deg_ls)
deg_mtx <- sapply(deg_ls, function(x) {
  as.numeric(all_deg %in% x)
}) %>% data.frame()
rownames(deg_mtx) <- all_deg
pdf(file = paste0(dir_out, "Fig4_venn_GFP_pos_vs_neg_", today, ".pdf"), width = 6, height = 6)
vennCounts(deg_mtx, include = "up") %>% vennDiagram(circle.col = tissue_colors()[c(2, 3, 4, 5, 1)])
dev.off()

tmp <- deg_mtx %>% rownames_to_column(var = "gene") %>% group_by(SPL, MLN, PP, SI, COL) %>%
  summarise(genes = paste(gene, collapse = "\n")) %>% 
  mutate(SPL_c = ifelse(SPL == 0, "", "SPL"),
         MLN_c = ifelse(MLN == 0, "", "MLN"),
         PP_c = ifelse(PP == 0, "", "PP"),
         SI_c = ifelse(SI == 0, "", "SI"),
         COL_c = ifelse(COL == 0, "", "COL"),
         tissues = paste(SPL_c, MLN_c, PP_c, SI_c, COL_c, sep = ","))
for (i in 1:nrow(tmp)) {
  cat("\n=============\n", tmp$tissues[i], "\n", sep = "")
  cat("***************\n", tmp$genes[i],"\n", sep = "")
}


```




## 7: The SLAMF6+ and CXCR6+ populations 

(1) COL DSS colitis data
see 1_code/colitis/analysis.Rmd

```{r}
dir_colitis <- "../../2_pipeline/colitis/"
pt.size <- 1
cont_markers <- read.table("../../0_data/gene_lists/Genes_for_cleanup.csv", stringsAsFactors = FALSE)[,1]
cont_markers <- union(cont_markers, c("Ptprc", "Tcrg-C4", "Tcrg-C2", "Tcrg-C1", "Il17a"))
genes_of_interest <- read.table("../../0_data/gene_lists/genes_of_interest_Alex_20190421.csv", stringsAsFactors = FALSE)[,1]
genes_of_interest <- union(genes_of_interest, "Il17a")
sigs_sc <- readRDS("../../2_pipeline/differential_expression/SPL_EAE_0_1/2020-07-22/sc_signature.rds")
names(sigs_sc)
names(sigs_sc) <- c("CXCR6+_sc", "SLAMF6+_sc")
names(sigs_sc) 
sigs_bulk <- readRDS("../../2_pipeline/spl_clusters/bulk/2020-02-04/signature_slamf6_vs_cxcr6_filtered.rds")
names(sigs_bulk)
names(sigs_bulk) <- c("SLAMF6+_bulk", "CXCR6+_bulk")
names(sigs_bulk)
sigs_ls <- c(sigs_sc[c("SLAMF6+_sc", "CXCR6+_sc")], sigs_bulk[c("SLAMF6+_bulk", "CXCR6+_bulk")])

trt <- "DSS"
so <- readRDS(file = paste0(dir_colitis, "so_COL_", trt, "_sct.rds"))

# remove contamination, Treg and proliferating clusters
# markers <- FindAllMarkers(so) # validated that Cluster 8 is contamination (Cd74+), C4 is Treg and T7 is proliferating.

so <- so[, !(so$seurat_clusters %in% c(4, 7, 8))]
so <- RunPCA(so)
ElbowPlot(so, ndims = 50)
so <- RunUMAP(so, reduction = "pca", dims = 1:30)

so <- FindNeighbors(so, dims = 1:30)
so <- FindClusters(so)
markers <- FindAllMarkers(so)
write.table(markers, file = paste0(dir_out, "cluster_markers_DSS.csv"), sep = ",", row.names = F)


p_bar <- so@meta.data %>%
  group_by(seurat_clusters, orig.ident) %>%
  tally() %>%
  ggplot(aes(x=seurat_clusters, y = n))+
  geom_bar(aes(fill = seurat_clusters), stat = "identity") +
  geom_text(aes(y = n, label = n), vjust=0) +
  facet_wrap(~orig.ident) +
  theme_bw() +
  ggtitle(paste0("Cluster sizes (Total = ", ncol(so), " cells)"))

p_umap_cluster <- DimPlot(object = so, reduction = "umap", group.by = "seurat_clusters", pt.size = pt.size, label = FALSE)

pdf(file = paste0(dir_out, "UMAP_contamination_", trt, ".pdf"), width = 12)
grid.arrange(p_bar, p_umap_cluster, nrow = 1)
for (g in cont_markers) {
  if (! (g %in% rownames(so))) {
    cat(g, ": not in data...\n"); next
  }
  cat(g, "\n")
  p1 <- VlnPlot(object = so, features = g, pt.size = pt.size)
  p2 <- FeaturePlot(object = so, features = g,
                    min.cutoff = "q5", reduction = "umap", pt.size = pt.size)
  grid.arrange(p1, p2, nrow = 1)
}
dev.off()
pdf(file = paste0(dir_out, "UMAP_genes_of_interest_", trt, ".pdf"), width = 12)
grid.arrange(p_bar, p_umap_cluster, nrow = 1)
for (g in genes_of_interest) {
  if (! (g %in% rownames(so))) {
    cat(g, ": not in data...\n"); next
  }
  cat(g, "\n")
  p1 <- VlnPlot(object = so, features = g, pt.size = pt.size)
  p2 <- FeaturePlot(object = so, features = g,
                    min.cutoff = "q5", reduction = "umap", pt.size = pt.size)
  grid.arrange(p1, p2, nrow = 1)
}
dev.off()

## signature scores
so <- AddModuleScore(so, features = sigs_ls, name = "SIGS")
colnames(so@meta.data)[grep("SIGS", colnames(so@meta.data))] <- names(sigs_ls)


plt_df <- data.frame(UMAP_1 = so@reductions$umap@cell.embeddings[,1], 
                     UMAP_2 = so@reductions$umap@cell.embeddings[,2],
                     slamf6_sc = so$`SLAMF6+_sc`,
                     cxcr6_sc = so$`CXCR6+_sc`,
                     slamf6_bulk = so$`SLAMF6+_bulk`,
                     cxcr6_bulk = so$`CXCR6+_bulk`) %>% 
  mutate(slamf6_sc_rank_norm = rank(slamf6_sc) / length(slamf6_sc),
         cxcr6_sc_rank_norm = rank(cxcr6_sc) / length(cxcr6_sc),
         slamf6_bulk_rank_norm = rank(slamf6_bulk) / length(slamf6_bulk),
         cxcr6_bulk_rank_norm = rank(cxcr6_bulk) / length(cxcr6_bulk)) 


# p_slamf6_sc <- plt_df %>% 
#   ggplot(.,  aes(x = UMAP_1, y = UMAP_2, color = slamf6_sc_rank_norm)) + 
#   geom_point(size = pt.size) +
#   scale_color_gradientn(colours = colorRamps::matlab.like2(10)) +
#   labs(color = "Normalized\nsignature\nscore", title = "SLAMF6+ Signature") +
#   theme_classic()
# p_cxcr6_sc <- plt_df %>% 
#   ggplot(.,  aes(x = UMAP_1, y = UMAP_2, color = cxcr6_sc_rank_norm)) + 
#   geom_point(size = pt.size) +
#   scale_color_gradientn(colours = colorRamps::matlab.like2(10)) +
#   labs(color = "Normalized\nsignature\nscore", title = "CXCR6+ Signature") +
#   theme_classic()


p_slamf6_bulk <- plt_df %>% 
  ggplot(.,  aes(x = UMAP_1, y = UMAP_2, color = slamf6_bulk_rank_norm)) + 
  geom_point(size = pt.size) +
  scale_color_gradientn(colours = colorRamps::matlab.like2(10)) +
  labs(color = "Normalized\nsignature\nscore", title = "SLAMF6+ Signature") +
  theme_classic()
p_cxcr6_bulk <- plt_df %>% 
  ggplot(.,  aes(x = UMAP_1, y = UMAP_2, color = cxcr6_bulk_rank_norm)) + 
  geom_point(size = pt.size) +
  scale_color_gradientn(colours = colorRamps::matlab.like2(10)) +
  labs(color = "Normalized\nsignature\nscore", title = "CXCR6+ Signature") +
  theme_classic()

# pdf(file = paste0(dir_out, "Fig7_UMAP_SLAMF6_CXCR6_SC_signatures_DSS_", today, ".pdf"), width = 9, height = 4)
# grid.arrange(p_slamf6_sc + theme(legend.position = "none"), 
#              p_cxcr6_sc + theme(legend.position = "none"),
#              nrow = 1, right = get_legend(p_slamf6_sc))
# dev.off()

pdf(file = paste0(dir_out, "Fig7_UMAP_SLAMF6_CXCR6_bulk_signatures_DSS_", today, ".pdf"), width = 9, height = 4)
grid.arrange(p_slamf6_bulk + theme(legend.position = "none"), 
             p_cxcr6_bulk + theme(legend.position = "none"), 
             nrow = 1, right = get_legend(p_slamf6_bulk))
dev.off()
```



(2) Smillie human COL colitis data (same as Fig.2)

Use CD4+ data.
```{r}
so <- readRDS("../../2_pipeline/human_smillie/so_imm_2021-01-23.rds")
DefaultAssay(so) <- "RNA"
table(so$Cluster, so$Health)
#                       Healthy Inflamed Non-inflamed
# CD4+ Activated Fos-hi    3524     2911         3950
# CD4+ Activated Fos-lo    3559     2796         3064
# CD4+ Memory              4992    12116         9657
# CD4+ PD1+                  75      294           48
so$Health <- factor(so$Health, levels = c("Healthy", "Non-inflamed", "Inflamed"))
# so <- so[,so$Health == "Inflamed"]
# so <- so[,so$Cluster %in% c("CD4+ Activated Fos-hi", "CD4+ Activated Fos-lo")]
# so <- so[,so$Cluster %in% c("CD4+ Memory")]
grep("CD8", rownames(so), value = T)
table(so[["RNA"]]@counts["CD8A",] > 0, so[["RNA"]]@counts["CD8B",] > 0) / ncol(so)
  #            FALSE       TRUE
  # FALSE 0.89882093 0.03890521
  # TRUE  0.03066871 0.03160516

so <- so[, (so[["RNA"]]@counts["CD8A",] + so[["RNA"]]@counts["CD8B",]) == 0] # exclude cells expressing CD8A or CD8B
so <- NormalizeData(so)

## compute CXCR6+ and SLAMF6+ signatures
# load and parse signature
sigs_ls <- readRDS("../../2_pipeline/spl_clusters/bulk/2020-02-04/signature_slamf6_vs_cxcr6_filtered.rds")
names(sigs_ls)
names(sigs_ls) <- c("SLAMF6+_bulk", "CXCR6+_bulk")
names(sigs_ls)
sapply(sigs_ls, length)
sigs_ls <- mouse_to_human(sigs_ls)
sapply(sigs_ls, length)
# compute signature scores
# so@meta.data <- so@meta.data[,-grep("SLAMF6", colnames(so@meta.data))]
# so@meta.data <- so@meta.data[,-grep("CXCR6", colnames(so@meta.data))]
so <- AddModuleScore(so, features = sigs_ls, name = "SIGS")
colnames(so@meta.data)[grep("SIGS", colnames(so@meta.data))] <- names(sigs_ls)

# violin plot
plt_df <- so@meta.data[,c("Health", "Cluster", "SLAMF6+_bulk", "CXCR6+_bulk")]
set.seed(1)
plt_df$show_point <- runif(nrow(plt_df)) < 0.1 # show jittered points for 10% of cells
p_SLAMF6 <- ggplot(plt_df, aes(x = Health, y = `SLAMF6+_bulk`, fill = Health)) +
  geom_violin() +
  geom_jitter(aes(shape = show_point), size = 0.01) +
  geom_boxplot(width = 0.1, fill = NA, color = "white", outlier.shape = NA) +
  facet_wrap(~Cluster, nrow = 1) +
  scale_shape_manual(values = c(NA, 19), guide = FALSE) +
  scale_fill_manual(values = c("Healthy" = "#3d5a80", "Non-inflamed" = "#8d99ae", "Inflamed" = "#98c1d9")) +
  labs(x = "", y = "Signature score", title = expression(SLAMF6^{textstyle("+")}~signature)) +
  theme_classic() +
  theme(legend.position = "none",
        axis.title = element_text(size = 15, color = "black"),
        axis.text.x = element_text(size = 12, color = "black", angle = 45, hjust = 1),
        axis.text.y = element_text(size = 12, color = "black"),
        plot.title = element_text(hjust = 0.5, face = "bold"))

p_CXCR6 <- ggplot(plt_df, aes(x = Health, y = `CXCR6+_bulk`, fill = Health)) +
  geom_violin() +
  geom_jitter(aes(shape = show_point), size = 0.01) +
  geom_boxplot(width = 0.1, fill = NA, color = "white", outlier.shape = NA) +
  facet_wrap(~Cluster, nrow = 1) +
  scale_shape_manual(values = c(NA, 19), guide = FALSE) +
  scale_fill_manual(values = c("Healthy" = "#3d5a80", "Non-inflamed" = "#8d99ae", "Inflamed" = "#98c1d9")) +
  labs(x = "", y = "Signature score", title = expression(CXCR6^{textstyle("+")}~signature)) +
  theme_classic() +
  theme(legend.position = "none",
        axis.title = element_text(size = 15, color = "black"),
        axis.text.x = element_text(size = 12, color = "black", angle = 45, hjust = 1),
        axis.text.y = element_text(size = 12, color = "black"),
        plot.title = element_text(hjust = 0.5, face = "bold"))
pdf(file = paste0(dir_out, "Fig7_violin_SLAMF6_CXCR6_bulk_signatures_Smillie_", today, ".pdf"), width = 14, height = 4)
grid.arrange(p_SLAMF6, p_CXCR6, nrow = 1)
dev.off()

## visualize signatures and genes on UMAPs
so <- FindVariableFeatures(so)
so <- ScaleData(so)
so <- RunPCA(so)
ElbowPlot(so, ndims = 50)
so <- RunUMAP(so, reduction = "pca", dims = 1:30, min.dist = 0.5)
DimPlot(so, group.by = "Cluster", reduction = "umap")

plt_df <- data.frame(UMAP_1 = so@reductions$umap@cell.embeddings[,1],
                     UMAP_2 = so@reductions$umap@cell.embeddings[,2],
                     label = str_wrap(so$Cluster, width = 80),
                     slamf6_expr = so[["RNA"]]@data["SLAMF6",],
                     cxcr6_expr = so[["RNA"]]@data["CXCR6",],
                     slamf6_bulk = so$`SLAMF6+_bulk`,
                     cxcr6_bulk = so$`CXCR6+_bulk`) %>%
  mutate(slamf6_bulk_rank_norm = rank(slamf6_bulk) / length(slamf6_bulk),
         cxcr6_bulk_rank_norm = rank(cxcr6_bulk) / length(cxcr6_bulk))
set.seed(1)
plt_df <- plt_df[sample(1:nrow(plt_df), nrow(plt_df), replace = F),]

pt_size <- 0.1
p_label <- plt_df %>%
  ggplot(.,  aes(x = UMAP_1, y = UMAP_2, color = label)) +
  geom_point(size = pt_size) +
  guides(colour = guide_legend(override.aes = list(size = 2))) +
  labs(color = "Cell label", title = "") +
  theme_classic()
p_slamf6_gene <- plt_df %>%
  arrange(slamf6_expr) %>% 
  ggplot(.,  aes(x = UMAP_1, y = UMAP_2, color = slamf6_expr)) +
  geom_point(size = pt_size) +
  scale_color_gradientn(colours = c("grey", "blue")) +
  labs(color = "Exp.", title = "SLAMF6+") +
  theme_classic()
p_cxcr6_gene <- plt_df %>%
  arrange(cxcr6_expr) %>%
  ggplot(.,  aes(x = UMAP_1, y = UMAP_2, color = cxcr6_expr)) +
  geom_point(size = pt_size) +
  scale_color_gradientn(colours = c("grey", "blue")) +
  labs(color = "Exp.", title = "CXCR6+") +
  theme_classic()
p_slamf6_bulk <- plt_df %>%
  ggplot(.,  aes(x = UMAP_1, y = UMAP_2, color = slamf6_bulk_rank_norm)) +
  geom_point(size = pt_size) +
  scale_color_gradientn(colours = colorRamps::matlab.like2(10)) +
  labs(color = "Normalized\nsignature score", title = "SLAMF6+ Signature") +
  theme_classic()
p_cxcr6_bulk <- plt_df %>%
  ggplot(.,  aes(x = UMAP_1, y = UMAP_2, color = cxcr6_bulk_rank_norm)) +
  geom_point(size = pt_size) +
  scale_color_gradientn(colours = colorRamps::matlab.like2(10)) +
  labs(color = "Normalized\nsignature score", title = "CXCR6+ Signature") +
  theme_classic()

pdf(file = paste0(dir_out, "Fig7_UMAP_SLAMF6_CXCR6_bulk_signatures_Smillie_Inflamed_", today, ".pdf"), width = 20, height = 7)
grid.arrange(
  p_label + theme(legend.position = "none"),
  get_legend(p_label),
  p_slamf6_gene + theme(legend.position = "none"),
  get_legend(p_slamf6_gene),
  p_slamf6_bulk + theme(legend.position = "none"),
  p_cxcr6_gene + theme(legend.position = "none"), 
  get_legend(p_cxcr6_gene),
  p_cxcr6_bulk + theme(legend.position = "none"),
  get_legend(p_slamf6_bulk),
  layout_matrix = matrix(c(1, 2, 3, 4, 5, 9, 1, 2, 6, 7, 8, 9), nrow = 2, byrow = T),
  widths = c(2, 0.5, 1, 0.5, 1, 0.5))
dev.off()






```


<!-- Use all data. -->
<!-- ```{r} -->
<!-- dat <- Matrix::readMM("../../0_data/other/Smillie_human_UC/SCP259/expression/5cdc540d328cee7a2efc234a/gene_sorted-Imm.matrix.mtx") -->
<!-- dat_gene <- read.table("../../0_data/other/Smillie_human_UC/SCP259/expression/5cdc540d328cee7a2efc234a/Imm.genes.tsv", stringsAsFactors = F) -->
<!-- dat_barcode <- read.table("../../0_data/other/Smillie_human_UC/SCP259/expression/5cdc540d328cee7a2efc234a/Imm.barcodes2.tsv", stringsAsFactors = F) -->
<!-- all_meta <- read.table("../../0_data/other/Smillie_human_UC/SCP259/metadata/all.meta2.txt", stringsAsFactors = F, sep = "\t", header = T) -->
<!-- rownames(dat) <- dat_gene$V1 -->
<!-- colnames(dat) <- dat_barcode$V1 -->
<!-- dat_meta <- all_meta %>%  -->
<!--   filter(NAME %in% dat_barcode$V1) %>%  -->
<!--   # filter(Health == "Healthy") %>%  -->
<!--   column_to_rownames(var = "NAME")  -->
<!-- table(dat_meta$Cluster) -->
<!-- dat <- dat[,rownames(dat_meta)] -->

<!-- tmp = table(dat["IL17A",]>0, dat_meta$Cluster)  -->
<!-- round(t(tmp) / colSums(tmp) * 100, 1) -->
<!-- # so <- CreateSeuratObject(counts = dat[,rownames(dat_meta)], meta.data = dat_meta) -->
<!-- # rm(dat, dat_gene, dat_barcode, all_meta) -->
<!-- ``` -->



## 9. Population RNA-seq of Il17aCre Il23rfl/fl CXCR6+ cells
see 1_code/il23rko/analysis.Rmd


PCA
```{r}
dir_il23rko <- "../../2_pipeline/Il23rko/"
cell_type <- "CXCR6+"
date_prep <- "2021-06-09"
dds <- readRDS(file = paste0(dir_il23rko, "dds_", date_prep, ".rds"))
dds_cxcr6_308 <- dds[,(dds$experiment == "As308") & (dds$cell_type == cell_type)]
vsd_cxcr6_308 <- vst(dds_cxcr6_308)
rv <- rowVars(assay(vsd_cxcr6_308))
select <- order(rv, decreasing=TRUE)[seq_len(min(1000, length(rv)))]
pca <- prcomp(t(assay(vsd_cxcr6_308)[select,]))
percentVar <- 100 * (pca$sdev ^ 2) / sum(pca$sdev ^ 2); names(percentVar) <- colnames(pca$x)

plt_df <- data.frame(PC_horiz = pca$x[, "PC1"], PC_vert = pca$x[, "PC2"], colData(vsd_cxcr6_308))
p <- ggplot(data = plt_df, aes(x = PC_horiz, y = PC_vert, color = toupper(condition))) + 
  geom_point(size = 3) + 
  scale_color_manual(name = "Condition", values = c("KO" = rgb(76, 169, 95, maxColorValue = 255), 
                                                    "WT" = rgb(88, 89, 91, maxColorValue = 255))) +
  labs(x = paste0("PC1 (", round(percentVar["PC1"]), "% variance)"),
       y = paste0("PC2 (", round(percentVar["PC2"]), "% variance)")) + 
  scale_x_continuous(expand = c(0.1, 0.1)) +
  scale_y_continuous(expand = c(0.1, 0.1)) +
  theme_bw()
  
pdf(file = paste0(dir_out, "Fig9_Il23rKO_PCA_As308_", today, ".pdf"), width = 5, height = 3.5)
p
dev.off()

```


Heatmap of DE genes.
```{r}
thresh_padj <- 0.05
date_de <- "2021-06-09"
res_edger_df <- list()
e <- "As308_only"
cell_type <- "CXCR6+"

res_edger <- readRDS(file = paste0(dir_il23rko, "de_edger_", cell_type, "_", e, "_", date_de, ".rds"))
write.xlsx(res_edger$table[order(res_edger$table$PValue),], 
           file = paste0(dir_out, "Table_S6.xlsx"), 
           sheetName = "Il23rko vs. wt",
           rowNames = T, colNames = T, overwrite = T)

plt_df <- as.data.frame(res_edger) %>%
    rownames_to_column(var = "gene") %>%
    filter(!is.na(PValue)) %>%
    mutate(up_in = ifelse(FDR >= thresh_padj, "NS", ifelse(logFC < 0, "WT", "KO")) %>% factor(., levels = c("KO", "WT", "NS")))
expr <- assay(vsd_cxcr6_308)[plt_df$gene[plt_df$up_in != "NS"],] %>% t %>% scale %>% t
hc_row <- hclust(dist(expr), method = "complete")
expr <- expr[hc_row$order,]
# hc_col <- hclust(dist(t(expr)), method = "complete")
# expr <- expr[,hc_col$order]

## heatmap
ann_col <- data.frame(
  Condition = toupper(as.character(colData(dds_cxcr6_308)[,"condition"]))
)
rownames(ann_col) <- colnames(dds_cxcr6_308)
ann_colors <- list(Condition = c("WT" = "black", "KO" = rgb(76, 169, 95, maxColorValue = 255)))
p <- pheatmap(expr,
         cluster_cols = F, cluster_rows = F,
         color = colorRampPalette(c("blue", "white", "red"))(100),
         breaks = seq(-2, 2, length.out = 101),
         annotation_col = ann_col,
         annotation_colors = ann_colors,
         show_rownames = T,
         show_colnames = F,
         main = "")
which_row_names <- which(p$gtable$layout$name == "row_names")
# p$gtable$grobs[[which_row_names]]$rot = 0
# p$gtable$grobs[[which_row_names]]$hjust = 0
# p$gtable$grobs[[which_row_names]]$vjust = 1
p$gtable$grobs[[which_row_names]]$gp$fontsize = 3
p$gtable$grobs[[which_row_names]]$gp$font = 3L

pdf(paste0(dir_out, "Fig_heatmap_DEG_Il23rKO_", today, ".pdf"), width = 5, height = 7)
grid.arrange(p$gtable)
dev.off()

pdf(paste0(dir_out, "Fig9_heatmap_DEG_Il23rKO_no_gene_name_", today, ".pdf"), width = 5, height = 7)
tmp <- p
tmp$gtable$grobs[[which_row_names]]$gp$fontsize = 0
grid.arrange(tmp$gtable)
dev.off()
```


GSEA barplot.
```{r}
date_fgsea <- "2021-07-29"
test_out <- readRDS(file = paste0(dir_il23rko, "fGSEA_", cell_type, "_", date_fgsea, ".rds"))
sigs_to_plot <- c(
  'Th17_Th1_like_memory-minus' = 'Th17/Th1-like memory\n(minus)',
  'Th17_Th1_like_effector_CNS-minus' = 'Th17/Th1-like effector in CNS\n(minus)',
  'Th17_Th1_like_effector-minus' = 'Th17/Th1-like effector\n(minus)',
  'Th17_self_renewing-plus' = 'Th17 self-renewing',
  'Pathogenic_Th17' = 'Pathogenic Th17',
  '_x0009_HALLMARK_INTERFERON_GAMMA_RESP' = 'HALLMARK interferon\ngamma response'
)

## make bar plot
p <- test_out %>% 
  # filter(pval < 0.2) %>%
  filter(pathway %in% names(sigs_to_plot)) %>%
  mutate(signature = sigs_to_plot[pathway]) %>%
  # mutate(signature = pathway) %>% 
  mutate(enriched_in = ifelse(NES > 0, "KO", "WT")) %>%
  ggplot(aes(x = reorder(signature, NES), y = NES)) +
  geom_col(aes(fill = enriched_in), width = 0.7) +
  geom_hline(yintercept = 0, color = "black") +
  scale_fill_manual(values = c("KO" = rgb(76, 169, 95, maxColorValue = 255), 
                               "WT" = "black")) +
  coord_flip() +
  labs(x="", y="Normalized Enrichment Score", fill = "Enriched in") +
  theme_bw() +
  theme(legend.title = element_text(size = 12),
        legend.text = element_text(size = 12),
        legend.box.margin=margin(-10,-10,-10,-10),
        axis.title = element_text(size = 12),
        axis.text = element_text(size = 10, color = "black"))
p
pdf(paste0(dir_out, "Fig_Il23rKO_barplot_enrichment_", cell_type, "_", today, ".pdf"), width = 6, height = 4)
print(p)
dev.off()

pdf(paste0(dir_out, "Fig_Il23rKO_barplot_enrichment_", cell_type, "_tex_on_right_", today, ".pdf"), width = 6, height = 4)
print(p + scale_x_discrete(position = "top"))
dev.off()

```

<!-- Load data. -->
<!-- ```{r} -->
<!-- prep_type_date <- "dominant_TCR_2020-03-24" -->
<!-- so_all <- readRDS(paste0(dir_proj, "2_pipeline/preprocessing/so_processed_", prep_type_date, ".rds")) -->
<!-- date_cluster <- "2020-04-13" -->
<!-- meta <- read.table(paste0(dir_proj, "2_pipeline/clustering/UT_GFPpos_inter/",  -->
<!--                           date_cluster, "/FILES/meta_data.txt"),  -->
<!--                    header = T, sep = " ", row.names = 1) -->
<!-- reductions <- readRDS(paste0(dir_proj, "2_pipeline/clustering/UT_GFPpos_inter/", date_cluster, "/FILES/reductions.rds")) -->
<!-- stopifnot(all(rownames(meta) == rownames(reductions$umap@cell.embeddings))) -->

<!-- so <- so_all[,rownames(meta)] ## restrict to SPL EAE cells -->
<!-- so@meta.data <- meta -->
<!-- so$tissue_cluster <- paste(so$tissue, so$seurat_clusters, sep = "_") -->
<!-- so@reductions <- reductions -->

<!-- so <- NormalizeData(so, assay = "RNA", scale.factor = 10^6) ## compute log(CPM+1) -->
<!-- ``` -->